File: SEGUNDO/PDOO/PRACTICAS/PRACTICA2/java/deepSpace/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author joarpe
 */
public class SpaceStation implements SpaceFighter {
    private static final int MAXFUEL = 100;
    private static final double SHIELDLOSSPERUNITSHOT = 0.1;
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    private Damage pendingDamage = null;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar = null;
    // PRUEBA P5 EJEMPLO
    /*private boolean fastidia;
    private boolean spaceCity;*/
    
    private void assignFuelValue (float f) {
        if (f <= MAXFUEL) {
            fuelUnits = f;
        }
    }
    
    private void cleanPendingDamage () {
        if (pendingDamage.hasNoEffect()) {
            pendingDamage = null;
        }
    }
    
    SpaceStation (String n, SuppliesPackage supplies) {
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        nMedals = 0;
        weapons = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
        // PRUEBA P5 EJEMPLO
        //fastidia = false;
    }
    
    public SpaceStation(SpaceStation station) {
        name = station.getName();
        ammoPower=station.getAmmoPower();
        fuelUnits=station.getFuelUnits();
        shieldPower=station.getShieldPower();
        nMedals=station.getNMedals();
        weapons=station.getWeapons();
        shieldBoosters=station.getShieldBoosters();
        // PRUEBA P5 EJEMPLO
        //fastidia = false;
    }
    
    public void cleanUpMountedItems () {
        for (int i = 0; i < weapons.size(); ++i) {
            if (weapons.get(i).getUses() == 0) {
                weapons.remove(i);
                i--;
            }
        }
        
        for (int i = 0; i < shieldBoosters.size(); ++i) {
            if (shieldBoosters.get(i).getUses() == 0) {
                shieldBoosters.remove(i);
                i--;
            }
        }
    }
    
    public void discardHangar () {
        hangar = null;
    }
    
    public void discardShieldBooster (int i) {
        int size = shieldBoosters.size();
        
        if(i>=0 && i<size) {
            shieldBoosters.remove(i);
            
            if(pendingDamage!=null) {
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }
    
    public void discardShieldBoosterInHangar (int i) {
        if (hangar != null) {
            hangar.removeShieldBooster(i);
        }
    }
    
    public void discardWeapon (int i) {
        int size = weapons.size();
        
        if(i>=0 && i<size) {
            Weapon w = weapons.remove(i);
            
            if(pendingDamage!=null) {
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    
    public void discardWeaponInHangar (int i) {
        if (hangar != null) {
            hangar.removeWeapon(i);
        }
    }
    
    public float fire () {
        int size=weapons.size();
        float factor=1.0f;
        
        for (int i=0; i<size; ++i) {
            Weapon w = weapons.get(i);
            factor *= w.useIt();
        }
        
        return ammoPower*factor;
    }
    
    
    public float getAmmoPower () {
        return ammoPower;
    }
    
    public float getFuelUnits () {
        return fuelUnits;
    }
    
    public Hangar getHangar () {
        return hangar;
    }
    
    public String getName () {
        return name;
    }
    
    public int getNMedals () {
        return nMedals;
    }
    
    public Damage getPendingDamage () {
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters () {
        return shieldBoosters;
    }
    
    public float getShieldPower () {
        return shieldPower;
    }
    
    public float getSpeed () {
        float f;
        
        f = fuelUnits / MAXFUEL;
        
        return f;
    }
    
    public SpaceStationToUI getUIversion () {
        return (new SpaceStationToUI(this));
    }
    
    public ArrayList<Weapon> getWeapons () {
        return weapons;
    }
    
    public void mountShieldBooster (int i) {     
        if (hangar != null) {
            ShieldBooster aux = hangar.removeShieldBooster(i);
            
            if (aux != null) {
                shieldBoosters.add(aux);
            }
        }
    }
    
    public void mountWeapon (int i) {
        if (hangar != null) {
            Weapon aux = hangar.removeWeapon(i);
            
            if (aux != null) {
                weapons.add(aux);
            }
        }
    }
    
    public void move () {
        fuelUnits -= getSpeed();
        
        if (fuelUnits < 0) {
            fuelUnits = 0;
        }
    }
    
    public float protection () {
        int size=shieldBoosters.size();
        float factor= 1.0f;
        
        for (int i=0; i<size; ++i) {
            ShieldBooster s = shieldBoosters.get(i);
            factor*=s.useIt();
        }
        
        return factor*shieldPower;
    }
    
    public void receiveHangar (Hangar h) {
        if (hangar == null) {
            hangar = h;
        }
    }
    
    public boolean receiveShieldBooster (ShieldBooster s) {
        if (hangar != null) {
            return hangar.addShieldBoosters(s);
        }
        return false;
    }
    
    public ShotResult receiveShot (float shot) {
        float myProtection = protection();
        
        if (myProtection>=shot) {
            shieldPower-=SHIELDLOSSPERUNITSHOT*shot;
            if (shieldPower<0.0f)
                shieldPower=0.0f;
            return ShotResult.RESIST;
        }
        else {
            shieldPower=0.0f;
            return ShotResult.DONOTRESIST;
        }
    }
    
    public void receiveSupplies (SuppliesPackage s) {
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        fuelUnits += s.getFuelUnits();
    }
    
    public boolean receiveWeapon (Weapon w) {
        if (hangar != null) {
            return hangar.addWeapon(w);
        }
        return false;
    }
    
    public Transformation setLoot (Loot loot) {
        CardDealer dealer = CardDealer.getInstance();
        int h=loot.getNHangars();
        if (h>0) {
            hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        int elements = loot.getNSupplies();
        
        for (int i=0; i<elements; ++i) {
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        elements = loot.getNWeapons();
        
        for (int i=0; i<elements; ++i) {
            Weapon weap = dealer.nextWeapon();
            boolean received = receiveWeapon(weap);
        }
        elements=loot.getNShields();
        
        for(int i=0; i<elements; ++i) {
            ShieldBooster sh=dealer.nextShieldBooster();
            boolean received=receiveShieldBooster(sh);
        }
        int medals = loot.getNMedals();
        
        nMedals+=medals;
        
        if (loot.getEfficient())
            return Transformation.GETEFFICIENT;
        else if (loot.spaceCity())
            return Transformation.SPACECITY;
        else
            return Transformation.NOTRANSFORM;
    }
    
    public void setPendingDamage (Damage d) {
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState () {
        return ((pendingDamage == null) || (pendingDamage.hasNoEffect()));
    }
    
    public String toString () {
        return getUIversion().toString();
    }
    
    // PRUEBA P5 EJEMPLO
    /*public void setFastidia(boolean b) {
        fastidia = b;
    }
    public boolean getFastidia(){
        return fastidia;
    }
    public void setSpaceCity (boolean b) {
        spaceCity = b;
    }
    public boolean getSpaceCity () {
        return spaceCity;
    }*/
}

File: SEGUNDO/PDOO/PRACTICA4/ruby4/Ruby/SpaceStation.rb
#encoding:utf-8

require 'singleton'

require_relative 'SpaceStationToUI.rb'
require_relative 'ShotResult.rb'
require_relative 'CardDealer.rb'
require_relative 'SuppliesPackage.rb'
require_relative 'Hangar.rb'
require_relative 'Transformation.rb'

module Deepspace
    class SpaceStation 
        @@MAXFUEL = 100
        @@SHIELDLOSSPERUNITSHOT = 0.1

        private
        def assignFuelValue (f)
            if (f<=@@MAXFUEL)
                @fuelUnits = f 
            else
                @fuelUnits = @@MAXFUEL
            end
        end

        def cleanPendingDamage
            if (@pendingDamage != nil)
                if (@pendingDamage.hasNoEffect)
                    @pendingDamage = nil
                end
            end
        end

        public 
        def initialize (n, supplies)
            @name = n
            @ammoPower=0
            @fuelUnits=0
            @shieldPower=0
            receiveSupplies(supplies)
            @weapons = Array.new
            @shieldBoosters=Array.new
            @hangar = nil
            @pendingDamage= nil
            @nMedals = 0
        end

        def self.newCopy (station)
            aux = new(station.name, SuppliesPackage.new(station.ammoPower, station.fuelUnits, station.shieldPower))
            aux.nMedals = station.nMedals
            aux.weapons = Array.new(station.weapons)
            aux.shieldBoosters = Array.new(station.shieldBoosters)

            if (station.pendingDamage != nil)
                aux.pendingDamage = station.pendingDamage.copy
            end

            if (station.hangar != nil)
                aux.hangar = Hangar.newCopy(station.hangar)
            end

            return aux
        end

        def cleanUpMountedItems
            for i in(0..(@weapons.size-1)) do
                if (@weapons.at(i).uses==0)
                    @weapons.delete_at(i)
                    i=i-1
                end
            end

            for i in(0..(@shieldBoosters.size-1)) do
                if (@shieldBoosters.at(i).uses==0)
                    @shieldBoosters.delete_at(i)
                    i=i-1
                end
            end
        end

        def discardHangar
            @hangar = nil
        end

        def discardShieldBooster(i)
            size = @shieldBoosters.size()

            if ((i >= 0) && (i < size))
                @shieldBoosters.delete_at(i)

                if (@pendingDamage != nil)
                    @pendingDamage.discardShieldBooster
                    cleanPendingDamage()
                end
            end
        end

        def discardShieldBoosterInHangar(i)
            if (@hangar!=nil)
                @hangar.removeShieldBoosters(i)
            end
        end

        def discardWeapon(i)
            size = @weapons.size()

            if ((i >= 0) && (i < size))
                w = @weapons.delete_at(i)

                if (@pendingDamage != nil)
                    @pendingDamage.discardWeapon(w)
                    cleanPendingDamage()
                end
            end
        end

        def discardWeaponInHangar(i)
            if (@hangar != nil)
                @hangar.removeWeapons(i)
            end
        end

        def fire
            size = @weapons.size()
            factor = 1.0

            i = 0
            while (i < size)
                w = @weapons.at(i)
                factor = factor * w.useIt()
            
                i = i+1
            end

            return (@ammoPower*factor)
        end

        def ammoPower
            @ammoPower
        end

        def fuelUnits
            @fuelUnits
        end

        def hangar
            @hangar
        end

        def name
            @name
        end

        def nMedals
            @nMedals
        end

        def pendingDamage
            @pendingDamage
        end

        def shieldBoosters
            @shieldBoosters
        end

        def shieldPower
            @shieldPower
        end

        def getSpeed 
            @fuelUnits.to_f/@@MAXFUEL
        end

        def getUIversion
            SpaceStationToUI.new(self)
        end

        def weapons
            @weapons
        end

        def mountShieldBooster (i)
            if (@hangar!=nil)
                aux = @hangar.removeShieldBoosters(i)

                if (aux != nil) 
                    @shieldBoosters.push(aux)
                end
            end
        end

        def mountWeapon(i)
            if (@hangar!=nil)
                aux = @hangar.removeWeapons(i)
                
                if (aux != nil)
                    @weapons.push(aux)
                end
            end
        end

        def move
            @fuelUnits = @fuelUnits - getSpeed
            if (@fuelUnits<0)
                @fuelUnits=0
            end
        end

        def protection 
            size = @shieldBoosters.size()
            factor = 1.0

            i = 0
            while (i < size)
                s = @shieldBoosters.at(i)
                factor = factor * s.useIt()
            
                i = i+1
            end

            return (@shieldPower * factor)
        end

        def receiveHangar(h)
            if (@hangar == nil)
                @hangar=h 
            end
        end

        def receiveShieldBooster(s)
            if (@hangar!=nil)
                @hangar.addShieldBooster(s)
            else
                return false
            end
        end

        def receiveShot(shot)
            myProtection = protection()

            if (myProtection >= shot)
                @shieldPower = @shieldPower - (@@SHIELDLOSSPERUNITSHOT * shot)

                if (0.0 > @shieldPower)
                    @shieldPower = 0.0
                end

                return ShotResult::RESIST
            else
                @shieldPower = 0.0

                return ShotResult::DONOTRESIST
            end
        end

        def receiveSupplies (s)
            assignFuelValue(@fuelUnits + s.fuelUnits)
            @ammoPower = @ammoPower + s.ammoPower
            @shieldPower = @shieldPower + s.shieldPower
        end

        def receiveWeapon (w)
            if (@hangar!=nil)
                @hangar.addWeapon(w)
            else
                return false
            end
        end

        def setLoot (loot)
            dealer = CardDealer.instance()
            h = loot.nHangars()

            if (h > 0)
                @hangar = dealer.nextHangar()
                receiveHangar(@hangar)
            end

            elements = loot.nSupplies()

            i = 0
            while (i < elements)
                sup = dealer.nextSuppliesPackage()
                receiveSupplies(sup)
                
                i = i+1
            end

            elements = loot.nWeapons()

            i = 0
            while (i < elements)
                weap = dealer.nextWeapon()
                receiveWeapon(weap)
                
                i = i+1
            end

            elements = loot.nShields()

            i = 0
            while (i < elements)
                sh = dealer.nextShieldBooster()
                receiveShieldBooster(sh)
                
                i = i+1
            end

            medals = loot.nMedals()

            @nMedals = @nMedals + medals

            if (loot.efficient)
                return Transformation::GETEFFICIENT
            else
                if (loot.spaceCity)
                    return Transformation::SPACECITY
                else
                    return Transformation::NOTRANSFORM
                end
            end
        end

        def setPendingDamage(d)
            if (d != nil)
                @pendingDamage=d.adjust(@weapons, @shieldBoosters)
            end
        end

        def validState
            if (@pendingDamage==nil) 
                return true
            else 
                return @pendingDamage.hasNoEffect
            end
        end

        def to_s
            getUIversion.to_s
        end

    end
end
